### ISSUE TITLE
Investigate feasibility of using logout workaround during session upleveling

### ISSUE THREAD

**Issue by patrickvinograd** [department-of-veterans-affairs/va.gov-team#4902]
Can we feasibly use IAM's PKMS logout between LOA1 and LOA3 authentication attempts in order to work around the limitations of the SSOe session stuff.

**Comment by ericbuckley** [comment]
I ran a test yesterday, checking to see if a SSOe logout before attempting to uplevel would prevent the user from having to login a second time.  I wasn’t able to run the test programmatically using the existing saml_logout_callback function, however I did

- put a breakpoint in the code right before the upleveling process started
-  opened up a new tab with https://pint.eauth.va.gov/pkmslogout to manually log the user out
- continued with the upleveling process

This however didn’t work out as we had hoped.  The upleveling process triggered a new login flow in the same way that forceAuthn was previously doing.  I’m going to inspect the SAML payloads tomorrow to get a better understanding of what happened, but I’m guessing that when I manually logged out using the pkmslogout url, it also forwarded a message to ID.ME to log the user out.

**Comment by ericbuckley** [comment]
I'm close to be able to verify this (and check with the IAM team to confirm), but I believe the SAML request that SSOe posts to id.me always has the "forceAuthn" value set to true.  Thus even if we logout of SSOe, but not id.me, the second trip would still force the user to re-enter their credentials if id.me sees a force authn.

**Comment by ericbuckley** [comment]
I was able to confirm that after a logout, SSOe still POSTs SAML AuthN request with a ForceAuthn attribute of true, so it will force id.me to reauthenticate the user.  We should probably follow up with the IAM team to see why this header value is always there.  Below is an example of the root element and its attributes.

```
<samlp:AuthnRequest xmlns:ds="http://www.w3.org/2000/09/xmldsig#"
                    xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion"
                    xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol"
                    Destination="https://api.idmelabs.com/saml/SingleSignOnService"
                    ForceAuthn="true"
                    ID="FIMREQ_b4e2c6bf-016f-1718-bebb-8f81be4ff43c"
                    IsPassive="false"
                    IssueInstant="2020-01-17T19:02:38Z"
                    Version="2.0"
                    >
```

**Comment by ambikaroos** [comment]
IAM removed ForceAuthN yesterday (1/22/20) but when Eric tested, he found that the Pint environment never redirected back to local host, so its not working.

**Comment by ambikaroos** [comment]
@ericbuckley is blocked due to permissions, made a request to DSVA for access

**Comment by ericbuckley** [comment]
I'm now unblocked, and confimed that the logout process before upleveling works as intended.  This of course may not be needed though depending on what we hear from the IAM team next week.

**Comment by ericbuckley** [comment]
I just ran a few tests against SSOe to see if running an uplevel works without logging out first.  One of the tests I ran did work as expected (vets.gov.user+262), the other one did not (vets.gov.user+9).  I believe the latter failed on the vets-api code running on my laptop because of a missing mocked MVI profile for the vets.gov.user+9 user.

I'm fairly confident that this is going to work and we can close this issue as there is no need to run a logout before upleveling now.
